File: backend/enrichment/domain_validation.py
"""Validate DNS hostnames for IOC domain lookups (ASCII + IDN/punycode)."""

import re
from urllib.parse import urlparse

# DNS label: 1-63 chars, alphanumeric + hyphen, no leading/trailing hyphen.
_LABEL = r"[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?"
_DOMAIN_ASCII_RE = re.compile(rf"^(?:{_LABEL}\.)+{_LABEL}$")
_IPV4_RE = re.compile(r"^(?:\d{1,3}\.){3}\d{1,3}$")


def is_valid_domain(host: str) -> bool:
    """Return True when host is a plausible DNS hostname (FQDN with 2+ labels)."""
    if not host:
        return False

    host = host.rstrip(".").lower()
    if ":" in host and not host.startswith("["):
        try:
            parsed = urlparse(f"http://{host}")
            if parsed.hostname:
                host = parsed.hostname
        except ValueError:
            host = host.split(":", 1)[0]
    if not host or len(host) > 253:
        return False

    try:
        ascii_host = host.encode("idna").decode("ascii")
    except (UnicodeError, UnicodeDecodeError):
        return False

    if len(ascii_host) > 253 or not _DOMAIN_ASCII_RE.match(ascii_host):
        return False

    if _IPV4_RE.match(ascii_host):
        octets = ascii_host.split(".")
        if all(part.isdigit() and 0 <= int(part) <= 255 for part in octets):
            return False

    return True

File: frontend/src/utils/domainValidation.js
/**
 * Validate DNS hostnames for IOC domain detection (ASCII + IDN/punycode).
 */

const IPV4_RE = /^(\d{1,3}\.){3}\d{1,3}$/
const DOMAIN_LABEL = '[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?'
export const DOMAIN_ASCII_RE = new RegExp(`^(?:${DOMAIN_LABEL}\\.)+${DOMAIN_LABEL}$`)

/** Global pattern for extracting ASCII/punycode domains from free text. */
export const DOMAIN_EXTRACT_RE = new RegExp(
  `\\b(?:${DOMAIN_LABEL}\\.)+(?:${DOMAIN_LABEL})\\b`,
  'gi',
)

function isIpv4(host) {
  if (!IPV4_RE.test(host)) return false
  return host.split('.').every(octet => {
    const n = Number(octet)
    return Number.isInteger(n) && n >= 0 && n <= 255
  })
}

/**
 * @param {string} host
 * @returns {boolean}
 */
export function isValidDomain(host) {
  if (!host) return false

  const trimmed = host.replace(/\.$/, '').toLowerCase()
  if (!trimmed || trimmed.length > 253) return false

  let asciiHost = trimmed
  try {
    asciiHost = new URL(`http://${trimmed}`).hostname
  } catch {
    return false
  }

  if (!asciiHost || asciiHost.length > 253) return false
  if (isIpv4(asciiHost)) return false
  return DOMAIN_ASCII_RE.test(asciiHost)
}
